File: src/main/java/org/goafabric/catalog/job/condition/ConditionImportConfiguration.java
package org.goafabric.catalog.job.condition;

import org.goafabric.catalog.job.JobCompletionListener;
import org.springframework.batch.core.Job;
import org.springframework.batch.core.Step;
import org.springframework.batch.core.job.builder.JobBuilder;
import org.springframework.batch.core.launch.support.RunIdIncrementer;
import org.springframework.batch.core.repository.JobRepository;
import org.springframework.batch.core.step.builder.StepBuilder;
import org.springframework.batch.item.ItemReader;
import org.springframework.batch.item.ItemWriter;
import org.springframework.batch.item.file.builder.FlatFileItemReaderBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.ClassPathResource;
import org.springframework.transaction.PlatformTransactionManager;

@Configuration
public class ConditionImportConfiguration {

    @Bean
    public Job ConditionEoJob(Step conditionStep, JobCompletionListener listener, JobRepository jobRepository) {
        return new JobBuilder("conditionJob", jobRepository)
                .incrementer(new RunIdIncrementer())
                .listener(listener).flow(conditionStep).end()
                .build();
    }

    @Bean(name = "conditionStep")
    public Step conditionStep(ItemReader<ConditionEo> conditionItemReader,
                           ItemWriter<ConditionEo> ConditionEoItemWriter,
                           JobRepository jobRepository,
                           PlatformTransactionManager ptm) {
        return new StepBuilder("conditionStep", jobRepository)
                .<ConditionEo, ConditionEo>chunk(2, ptm)
                .reader(conditionItemReader)
                .writer(ConditionEoItemWriter)
                .build();
    }

    @Bean
    public ItemReader<ConditionEo> conditionItemReader() {
        return new FlatFileItemReaderBuilder<ConditionEo>()
                .name("conditionItemReader")
                .resource(new ClassPathResource("catalogs/icd10.csv"))
                .delimited().delimiter(";")
                .names(new String[]{"code", "display", "shortname"})
                .fieldSetMapper(fieldSet ->
                        new ConditionEo(fieldSet.readString("code"), fieldSet.readString("display"), fieldSet.readString("shortname")))
                .build();

    }
    
    @Bean
    public ItemWriter<ConditionEo> conditionItemWriter(ConditionRepository repository) {
        return repository::saveAll;
    }

}

File: src/main/kotlin/org/goafabric/catalog/job/condition/ConditionImportConfiguration.kt
package org.goafabric.catalog.job.condition

import org.goafabric.catalog.job.JobCompletionListener
import org.goafabric.catalog.job.StepCompletionListener
import org.springframework.batch.core.job.Job
import org.springframework.batch.core.job.builder.JobBuilder
import org.springframework.batch.core.repository.JobRepository
import org.springframework.batch.core.step.Step
import org.springframework.batch.core.step.builder.StepBuilder
import org.springframework.batch.infrastructure.item.Chunk
import org.springframework.batch.infrastructure.item.ItemReader
import org.springframework.batch.infrastructure.item.ItemWriter
import org.springframework.batch.infrastructure.item.file.builder.FlatFileItemReaderBuilder
import org.springframework.batch.infrastructure.item.file.mapping.FieldSetMapper
import org.springframework.batch.infrastructure.item.file.transform.FieldSet
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.ClassPathResource

@Configuration
class ConditionImportConfiguration {
    @Bean
    fun conditionEoJob(conditionStep: Step, listener: JobCompletionListener?, jobRepository: JobRepository): Job? {
        return JobBuilder("conditionJob", jobRepository) //.incrementer(new RunIdIncrementer())
            .listener(listener).flow(conditionStep).end()
            .build()
    }

    @Bean(name = ["conditionStep"])
    fun conditionStep(
        conditionItemReader: ItemReader<ConditionEo>,
        conditionEoItemWriter: ItemWriter<ConditionEo>,
        jobRepository: JobRepository
    ): Step {
        return StepBuilder("conditionStep", jobRepository)
            .chunk<ConditionEo, ConditionEo>(2)
            .reader(conditionItemReader)
            .writer(conditionEoItemWriter)
            .listener(StepCompletionListener())
            .build()
    }

    @Bean
    fun conditionItemReader(): ItemReader<ConditionEo> {
        return FlatFileItemReaderBuilder<ConditionEo>()
            .name("conditionItemReader")
            .resource(ClassPathResource("catalogs/icd10.csv"))
            .delimited().delimiter(";")
            .names("code", "display", "shortname")
            .fieldSetMapper(FieldSetMapper { fieldSet: FieldSet? ->
                ConditionEo(
                    code = fieldSet!!.readString("code"),
                    display = fieldSet.readString("display"),
                    shortname = fieldSet.readString("shortname")
                )
            })
            .build()
    }

    @Bean
    fun conditionItemWriter(repository: ConditionRepository): ItemWriter<ConditionEo> {
        return ItemWriter { entities: Chunk<out ConditionEo> -> repository.saveAll(entities) }
    }
}
